CRAM spec updates; large!

jkbonfield opened the pull request:
This is a good thrash at the CRAM spec, fixes #353 and also takes a good stab at making the data series decode order explicit rather than guesswork.  (The tables were almost correct already, but it was never explicitly mentioned.)

It's now using algorithmicx.sty.  I could use use algorithm.sty for slightly better formatting, but it's GPL and for want of a single extra line of formatting I couldn't be bothered to figure out what that means to a text document! (TeX of course is actually a language, but what is the "program" here?)

lbergelson commented:
@jmthibault79 Relevant to your work.

jmthibault79 commented:
Thank you!  I will review and comment inline.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
 \hline
-3 & int & RI & ref id & reference sequence id, not used for single reference slices, 
-reserved for future multiref slices. \tabularnewline
+- & - & Read names & See section \ref{subsec:readnames}\tabularnewline
```
Reference is `names` below (line 1076) so this does not render

jkbonfield replied:
Resolving things I've done locally just for ease of tracking.  I'll push an update once I've completed my round of revisions.

Thanks for taking the time to review.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
+- & - & Auxiliary tags & See section \ref{subsec:tags}\tabularnewline
 \hline
-6 & int & RG & read group & the read group identifier\tabularnewline
+- & - & Sequences & See sections \ref{subsec:mapped}and \ref{subsec:unmapped}\tabularnewline
```
Needs a space after `{subsec:mapped}`

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
+\hline
+\textbf{Bit flag} & \textbf{Comment} & \textbf{Description}\tabularnewline
+\hline
+0x1 & ! 0x40 \&\& ! 0x80 & template having multiple segments in sequencing\tabularnewline
```
Could you add more to this comment?  I have a few possible interpretations for the correct behavior so I'd like that to be made explicit here.

* Reading
1. bit `0x1` is strictly ignored, but stored as-is for round-tripping equality.  Always use the `0x40` and `0x80` values as the source of truth.
1. store bit `0x1` as-is.  If it conflicts with `0x40` and `0x80`, use the `0x1` value.  
1. bit `0x1` should be validated against the calculation from the `0x40` and `0x80` values, and the record rejected if these do not match
1. bit `0x1` should be corrected (overwritten) by the calculation from the `0x40` and `0x80` values

* Writing
1. write out the `0x1` bit from the `0x40` and `0x80` values, regardless of what it was previously
1. write out the `0x1` bit as the original value, even if it conflicts with the `0x40` and `0x80` values

jkbonfield replied:
These came from the SAM specification.  I expect Vadim added them because he wanted the CRAM specification to be self contained, which is reasonable.  It's overly terse and I agree confusing.

I'll review them and see what the SAM spec has to say too.  Do we want to just lift the SAM text verbatim, along with all the bullet points elaborating on them?  I think probably we can find a way to add them to the description part of the text, perhaps with footnotes for further explanatory bits.

jkbonfield replied:
Looking at my C code, I see I store BF verbatim (this is inefficient perhaps).  During read I start with the stored value and then modify it in some manners.

- Add 0x1 if a mate pair is present in this slice.
- Add 0x8 if mate is in this slice and is unmapped
- Add 0x20 if mate is in this slice and reverse complemented.

In other words it copes with these bits being missing and will repopulate (or even correct them if wrong), but it does store them.  I'm unsure why I did it this way.  Maybe the Java code doesn't (or didn't) store all these flags, or perhaps it's related to earlier CRAM versions.  I recall one of them having some bits missing - CRAMv1 maybe - and always generating them.  This doesn't work well though when the mate-pair is outside this slice.

I'd need to check the Java code to see what that does.

jmthibault79 replied:
Setting 0x1 if it has a mate pair fits the description, but I'm not understanding what that has to do with the 0x40 and 0x80 flags.

jkbonfield replied:
I'm struggling to understand that myself too actually.  I'll look at the Java code - maybe there is a clue there.  Agreed it isn't clear.

jkbonfield replied:
I just deleted it. :-)

It makes no sense and seems to have no apparent relationship to how either of our implementations work.  I guess it's something lost in the mists of time.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
+0x4 &  & segment unmapped\tabularnewline
 \hline
-10 & int & TL & tag ids & tag ids *3\tabularnewline
+0x8 & calculated* or stored in the mate's info & next segment in temlate unmapped\tabularnewline
```
spelling: template

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
-and normal integer value in all other cases. 
 
-*2 See \emph{mate record} section.
+The following pseudocode describes the general process of decoding an entire CRAM record.
```
This is excellent, thank you.

jkbonfield replied:
Thanks. :)

I am aware the codec section could be rewritten using algorithmicx too for consistency, but I added it after having gone through them.  I think this is maybe a second round of updates.  There's a lot more to do and incremental improvements is the easiest way given how much of the spec still needs a refresh.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
+\textbf{Data series type} & \textbf{Data series name} & \textbf{Field} & \textbf{Description}\tabularnewline
 \hline
- & \textbf{Data series type} & \textbf{Data series name} & \textbf{Field} & \textbf{Description}\tabularnewline
+int & RI & ref id & reference sequence id (only used in multiref slices)\tabularnewline
```
Should this be set to a particular constant when unused - or ignored?

jmthibault79 replied:
Yes, I like that

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,138 +921,325 @@ \section{\textbf{Record structure}}
+\vskip 20pt
+\begin{algorithmic}[1]
+\Procedure{DecodePositions}{}
+\If{$slice.ref\_id = -2$}
```
Please indicate here that `-2` means multi-ref

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -1139,96 +1327,94 @@ \subsubsection*{Base substitution codes (BS data series)}
-\subsubsection*{}
-
-\subsubsection*{Next mate bit flags (MF data series)}
+\subsubsection*{Decoe mapped read pseudocode}
```
spelling: Decode

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -1708,36 +1894,42 @@ \subsubsection*{Introduction}
+\subsubsection{External: codec ID 1}
 
-\subsubsection*{Definition}
+Can encode types \textit{Byte}, \textit{Integer}.
```
+ Long

jkbonfield replied:
I've added Long to io_lib for my CRAMv4 prototype and I see it's in htsjdk too, but htslib has no support for external long.

This section is ambiguous though as it describes Int and Byte being acceptable and then mentions Long in the next sentence.  In practice CRAM v3 (and before) have no longs in any data series, with long restricted purely to ltf8 encoding within container and slice headers.

We should be consistent though.  For v3 I think just drop long from external as it's not a possibility.

jmthibault79 replied:
OK.  The HTSJDK implementation of External Long is broken anyway.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -1708,36 +1894,42 @@ \subsubsection*{Introduction}
 \hline
 \end{tabular}
 
+\subsubsection{Beta coding: codec ID 6}
```
Curious about the ordering: why not in ID order?

jkbonfield replied:
I've no idea to be honest!  I didn't reorder anything, so this is as per the original author.  I'll take a look, but I can see a benefit to having golomb-rice immediately next to golomb.

Maybe we can list the recommended ones in order and then the deprecated ones that none of us use (in their own order).   It's arguable whether deprecating them is a change of specification that goes above simple clarification, but the practical side of it is they don't exist in the wild anyway.

jkbonfield replied:
Fixed reordering.  Also shuffled some other things around in this appendix.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -1746,13 +1938,38 @@ \subsubsection*{Parameters}
-\subsubsection*{Gamma coding}
+\subsubsection*{Examples}
+
+If we have integer values in the range 10 to 15 inclusive, the largest value would traditionally need 4 bits, but with an offset of -10 we can hold values 0 to 5, using a fixed size of 3 bits.
```
Great example.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -1996,37 +2226,74 @@ \subsubsection*{Parameters}
+- Read code lengths from codec parameters.
+\end{itemize}
+
+\item Compute canconical huffman codewords from code lengths\footnote{\url{https://en.wikipedia.org/wiki/Canonical_Huffman_code}}.
```
Spelling: canonical

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -2035,22 +2302,31 @@ \subsubsection*{Parameters}
+The length is decoded first followed by the data, followed by the next length and data, and so on.
+
+This encoding can therefore be considered as a nested encoding, with each pair of nested encodings containing their own set of parameters.
+The byte stream for parameters of the BYTE\_ARRAY\_LEN encoding is therefore the concatenation of the length and value encoding parameters as described in section~\ref{subsec:writing-bytes}.
```
Could you expand on this a little further with an example?  I'll make an attempt here:

As an example, if we used BETA encoding for length and EXTERNAL encoding for values, the parameters would be [ 6 (BETA encoding ID), len(beta params), beta params, 1 (EXTERNAL encoding ID), len(external params), external params ]

jkbonfield replied:
Good idea.

I went with HUFFMAN + external.  This has the benefit of showing how constant sized values can still utilise the byte array len mechanism, but in an efficient manner that doesn't actually use any storage for length field.  (I use this for integer aux tags.)

jkbonfield commented:
Another round of updates.

Hopefully this fixes most of the concerns.  Thanks for your comments @jmthibault79

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,143 +922,363 @@ \section{\textbf{Record structure}}
+\item[a] Bit 0x4 is the only reliable place to tell whether the read is unmapped.  If 0x4 is set, no assumptions may be made about bits 0x2, 0x100 and 0x800.
+\item[b] For segments within the same slice.
+\item[c] Bits 0x40 and 0x80 reflect the read ordering within each template inherent in the sequencing technology used, which may be independent from the actual mapping orientation.
+If 0x40 and 0x80 are both set, the read is part of a linear template (one where the template sequence is expected to be in a linear order), but it is neither the first nor the last read.
```
This is counterintuitive so I want to confirm: "first" = true and "last" = true means it is neither first nor last?

Maybe this would be better represented by a two-bit flag 0xC0 with 4 meaningful values?

jkbonfield replied:
This comes directly from the SAM specification, which basically has identical text bar a bit of clarification on "linear template".

I don't want to confuse things too much for people who are familiar with SAM and while I agree there is potential for a better way to describe it, I think that ship has sailed.

jmthibault79 replied:
ok

jkbonfield commented on CRAMv3.tex:
```diff
@@ -920,143 +922,363 @@ \section{\textbf{Record structure}}
 other hand, data series that do not require binary coding can be stored separately 
 in external blocks with some other compression applied to them independently.
 
+As CRAM data series may be interleaved within the same blocks\footnote{Interleaving can sometimes provide better compression, however it also adds dependency between types of data meaning it is not possible to selectively decode one data series if it co-locates with another data series in the same block.} understanding the order in which CRAM data series must be decoded is vital.
```
No, we can use the same external block ID and interleave to external blocks too.  This can improve compression ratios too if we are putting data with similar frequency distributions in the same block or data with strong correlation between values.

Eg imagine we had one tag X0 which is number of mismatches in this read and X1 which is number of mismatches in this plus mate.  There will be a strong correlation between the value of X1 and the value of X0 (at the very least X1 >= X0), so putting both together would mean an order-1 entropy encoder will probably do a decent job.

In practice we don't do this any more, so it remains untapped potential.

jmthibault79 replied:
Interesting!  I didn't realize that, thanks.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -1319,440 +1529,347 @@ \subsubsection*{BAM index}
-\hline
-c & a & 4095 \\
+
+For example, the bytes specifying a BYTE\_ARRAY\_LEN encoding, including the codec and parameters, for a 16-bit X0 auxiliary tag (``X0C'') may use HUFFMAN encoding to specify the length (always 2 bytes) and an EXTERNAL encoding to store the value to an external block with ID 200.
```
Great!

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -600,55 +605,30 @@ \subsubsection*{Preservation map}
-\hline
-TL & encoding\texttt{<}int\texttt{>} & tag ids  & list of tag ids, see tag encoding 
-section\tabularnewline
+RN\footnote{Note this is decoded after MF if ...} & encoding\texttt{<}byte[ ]\texttt{>} & read names & read names\tabularnewline
```
I don't see this footnote in the rendered PDF.  Same problem for TL.

jkbonfield replied:
Thanks.  I've reformatted these using LaTeX threeparttable.

jmthibault79 commented:
Done with this round of review, with one additional question: should we deprecate Gamma and Subexponential encodings as well?  My (limited) understanding is that they're similar to Golomb and Golomb-Rice in lack of use.

jkbonfield commented:
I think I have seen them in use in earlier versions of CRAM, unlike the golomb ones which got ditched prior to 1.0.

For CRAM v4 I'd like to either remove all of these and just use external codecs, or perhaps better to elevate them to be ways of writing day into external codecs (ie give them all a block ID instead of hard-coded to CORE) and permit bit-packing plus the possibility of a subsequent entropy encoding step.

If we had a bit-wise arithmetic coder coupled to a bit-packing method it *may* be better.  I haven't explored it though.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -1319,440 +1529,347 @@ \subsubsection*{BAM index}
+\subsubsection*{Decoding}
 
+\begin{enumerate}
+\item Read $i$ in unary form, counting the number of leading 1s (prefix) in the codeword (discard the trailing 0 bit).
```
Can we make this `u` as well, since it's the same as above?

jkbonfield replied:
Done.

jmthibault79 commented on CRAMv3.tex:
```diff
@@ -920,143 +922,363 @@ \section{\textbf{Record structure}}
+\Else
+    \State $alignment\_position \gets$ \Call{ReadItem}{AP, Integer}
+\EndIf
+\State $read\_group \gets$ \Call{ReadItem}{AG, Integer}
```
typo: should be RG

jkbonfield replied:
Fixed, thanks.

jmthibault79 commented:
I'm trying to map HTSJDK DataSeries to these updates, and there are a few I'm not finding here.  Could you add them?  (Or explain why they don't belong here)

* MQ - Mapping Quality
* TC - Tag Count
* TN - Tag Name and Type
* TM and TV - Test Marks (maybe these are intentionally out of spec)

jkbonfield commented:
>     * MQ - Mapping Quality

MQ is an oversight, and now added (along with QS too for mapped reads).  It was in the pseudocode and the main table, but not the record description.

>     * TC - Tag Count
>     * TN - Tag Name and Type

TC and TN are the old CRAM 1.0 format way of doing aux tags.  They have been replaced by TL and the TD preservation map.

>     * TM and TV - Test Marks (maybe these are intentionally out of spec)

I think these are tests internal to the Java code.  They don't appear to exist in the files I've looked at.

Anyway I pushed a new copy of the PR now with the latest round of changes.

I also went for broke and added pseudo-code for the rANS decoder (I gave up at the encoder, but it's less important).  I have a noddy implementation to test it works (it didn't originally!) to check my description is correct.  It's hard to get right, but surprisingly simple when written down that way.

I think this is enough for now though.

jkbonfield commented:
@rbuels: would you please take a look at these CRAM spec modifications.  They are a lot of clarifications in codecs, record structure and the rANS decoder.  There is still much more to do for clarity - all the early sections haven't been touched.

Apart from labelling some codecs as deprecated because none of our code generates data that way, none of this is actually changing the file format in any way.  The intent is simply to clarify and in some cases to fix mistakes in the specification.

[CRAMv3.pdf](https://github.com/samtools/hts-specs/files/2600934/CRAMv3.pdf)

rbuels commented:
Some very nice improvements here! The pseudocode sections in particular are going to be very useful for future implementors of the spec.

rbuels commented:
I also am very fond of the explicit deprecations, those could save somebody a lot of time!

jkbonfield commented:
Thanks for the comments all.  Sorry @rbuels these all came too late for your implementation!  Was there anything specific you found incomprehensible that you had to refer to the source code to work out? (Probably lots, sadly.)

The work above is incomplete as there is much to of the spec to refresh, but I propose merging it after our next conference call as a step-wise improvement.  Part 1 of several, hopefully.
